pytorch_lightning/callbacks/swa.py

```diff
@@ -0,0 +1,206 @@
+
+        if isinstance(avg_fn, Callable):
+            raise MisconfigurationException("avg_fn should be function.")
+
+        self._swa_start = swa_epoch_start
+        self._swa_lrs = swa_lrs
+        self._annealing_epochs = annealing_epochs
+        self._annealing_strategy = annealing_strategy
+        self._avg_fn = avg_fn
+        self._device = device
+        self._model_contains_batch_norm = None
+
+    @property
+    def swa_model(self):
+        return getattr(self, "_average_model")
+
+    @staticmethod
+    def pl_module_contains_batch_norm(pl_module):
+        for module in pl_module.modules():
+            if isinstance(module, nn.modules.batchnorm._BatchNorm):
+                return True
+        return False
+
+    def reset_batch_norm_and_save_state(self, average_model, device):
+        self.momenta = {}
+        for module in average_model.modules():
+            if isinstance(module, nn.modules.batchnorm._BatchNorm):
+                running_mean_dtype = module.running_mean.dtype
+                running_var_dype = module.running_var.dtype
+                module.running_mean = torch.zeros_like(module.running_mean, device=device, dtype=running_mean_dtype)
+                module.running_var = torch.ones_like(module.running_var, device=device, dtype=running_var_dype)
+                self.momenta[module] = module.momentum
+                module.momentum = None
+                module.num_batches_tracked *= 0
+
+    def apply_momemta(self):
+        for bn_module in self.momenta.keys():
+            bn_module.momentum = self.momenta[bn_module]
+
+    def on_fit_start(self, trainer, pl_module):
+        self._average_model = LightningAveragedModel(pl_module, device=self._device, avg_fn=self._avg_fn)
+        optimizers = trainer.optimizers
+        lr_schedulers = trainer.lr_schedulers
+
+        if len(optimizers) > 1:
+            raise MisconfigurationException("SWA currently not supported for more than 1 optimizer.")
+
+        if len(lr_schedulers) > 1:
+            raise MisconfigurationException("SWA currently not supported for more than 1 lr_scheduler.")
+
+        self._max_epochs = trainer.max_epochs
+
+        self._swa_scheduler = SWALR(
+            optimizers[0],
+            swa_lr=self._swa_lrs,
+            anneal_epochs=self._annealing_epochs,
+            anneal_strategy=self._annealing_strategy,
+            last_epoch=self._max_epochs if self._annealing_strategy == "cos" else -1
+        )
+
+        self._model_contains_batch_norm = self.pl_module_contains_batch_norm(pl_module)
+
+        if self._model_contains_batch_norm:
+            # virtually increase max_epochs to perform batch norm update
+            trainer.max_epochs += 1
+
+    def on_train_epoch_start(self, trainer, pl_module):
+        if trainer.current_epoch == self._swa_start:
+            optimizers = trainer.optimizers
+            lr_scheduler = trainer.lr_schedulers[0]["scheduler"]
+
+            swa_scheduler = SWALR(
```

Comment:
we cannot use `self._swa_scheduler` here?

Reply:
Yes.